File: scripts/monitor-openai-call.py
#!/usr/bin/env python3
"""Sideband WebSocket monitor for an active OpenAI Realtime SIP call (Bearer auth)."""
import json
import os
import sys
import time

try:
    import websocket
except ImportError:
    print("pip install websocket-client", file=sys.stderr)
    sys.exit(1)

def load_key():
    key = os.environ.get("OPENAI_API_KEY", "")
    if key:
        return key.strip()
    env_path = os.path.join(os.path.dirname(__file__), "..", ".env")
    if os.path.isfile(env_path):
        for line in open(env_path):
            if line.startswith("OPENAI_API_KEY="):
                return line.split("=", 1)[1].strip()
    return ""

def main():
    if len(sys.argv) < 2:
        print(f"Usage: {sys.argv[0]} rtc_u1_...", file=sys.stderr)
        sys.exit(1)
    call_id = sys.argv[1]
    api_key = load_key()
    if not api_key:
        print("OPENAI_API_KEY not set", file=sys.stderr)
        sys.exit(1)

    url = f"wss://api.openai.com/v1/realtime?call_id={call_id}"
    print(f"Connecting sideband WS for {call_id}...")

    events = []

    def on_open(ws):
        print("WS open")
        ws.send(json.dumps({
            "type": "response.create",
            "response": {
                "instructions": "Greet briefly, then keep the conversation going. Ask the caller a question."
            },
        }))

    def on_message(ws, message):
        try:
            msg = json.loads(message)
            t = msg.get("type", "")
            if t.startswith("response.") or "transcription" in t or t.startswith("input_audio"):
                print(f"EVT {t}: {json.dumps(msg)[:300]}")
            events.append(t)
        except Exception:
            print(message[:200])

    def on_error(ws, error):
        print(f"WS error: {error}")

    def on_close(ws, code, reason):
        print(f"WS closed code={code} reason={reason}")

    ws = websocket.WebSocketApp(
        url,
        header=[f"Authorization: Bearer {api_key}"],
        on_open=on_open,
        on_message=on_message,
        on_error=on_error,
        on_close=on_close,
    )
    # Block for up to 90s
    end = time.time() + 90
    import threading
    t = threading.Thread(target=ws.run_forever, kwargs={"ping_interval": 20, "ping_timeout": 10})
    t.daemon = True
    t.start()
    while t.is_alive() and time.time() < end:
        time.sleep(1)
    try:
        ws.close()
    except Exception:
        pass
    print(f"Done. event_types={len(events)}")

if __name__ == "__main__":
    main()

File: scripts/monitor-openai-call.mjs
#!/usr/bin/env node
/**
 * Sideband OpenAI Realtime monitor (Bearer auth — works outside Workers).
 * Usage: node scripts/monitor-openai-call.mjs rtc_u1_...
 */
import { readFileSync } from "fs";
import { createRequire } from "module";
import { dirname, join } from "path";
import { fileURLToPath } from "url";

const require = createRequire(import.meta.url);
const WebSocket = require(join(dirname(fileURLToPath(import.meta.url)), "../bridge-worker/node_modules/ws"));

const callId = process.argv[2];
if (!callId) {
  console.error("Usage: monitor-openai-call.mjs <call_id>");
  process.exit(1);
}

function loadKey() {
  if (process.env.OPENAI_API_KEY) return process.env.OPENAI_API_KEY.trim();
  const env = readFileSync(join(dirname(fileURLToPath(import.meta.url)), "..", ".env"), "utf8");
  const m = env.match(/^OPENAI_API_KEY=(.+)$/m);
  return m ? m[1].trim() : "";
}

const apiKey = loadKey();
if (!apiKey) {
  console.error("OPENAI_API_KEY missing");
  process.exit(1);
}

const url = `wss://api.openai.com/v1/realtime?call_id=${encodeURIComponent(callId)}`;
console.log(`Connecting ${callId}...`);

const ws = new WebSocket(url, {
  headers: { Authorization: `Bearer ${apiKey}` },
});

ws.on("open", () => {
  console.log("WS open");
  const prompt = () => {
    ws.send(JSON.stringify({
      type: "response.create",
      response: {
        instructions:
          "Greet the caller, then ask them to say something. Keep the call alive with short responses. Do not hang up.",
      },
    }));
  };
  prompt();
  const keepalive = setInterval(() => {
    if (ws.readyState !== WebSocket.OPEN) { clearInterval(keepalive); return; }
    prompt();
  }, 12_000);
  ws.on("close", () => clearInterval(keepalive));
});

ws.on("message", (data) => {
  try {
    const msg = JSON.parse(data.toString());
    const t = msg.type || "";
    if (
      t.startsWith("response.audio") ||
      t.includes("transcription") ||
      t.startsWith("input_audio") ||
      t === "session.updated"
    ) {
      console.log(`EVT ${t}: ${JSON.stringify(msg).slice(0, 400)}`);
    }
  } catch (_) {}
});

ws.on("error", (e) => console.error("WS error:", e.message));
ws.on("close", (code, reason) => console.log(`WS closed ${code} ${reason}`));

setTimeout(() => {
  console.log("Monitor timeout 90s");
  ws.close();
  process.exit(0);
}, 90_000);
